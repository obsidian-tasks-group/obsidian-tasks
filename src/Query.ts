import chrono from 'chrono-node';
import { LayoutOptions } from './LayoutOptions';

import { Status, Task } from './Task';

type Sorting = 'status' | 'due' | 'done' | 'path' | 'description';

export class Query {
    private _limit: number | undefined = undefined;
    private _layoutOptions: LayoutOptions = new LayoutOptions();
    private _filters: ((task: Task) => boolean)[] = [];
    private _error: string | undefined = undefined;
    private _sorting: Sorting[] = [];

    private readonly noDueString = 'no due date';
    private readonly dueRegexp = /^due (before|after|on)? ?(.*)/;

    private readonly doneString = 'done';
    private readonly notDoneString = 'not done';
    private readonly doneRegexp = /^done (before|after|on)? ?(.*)/;

    private readonly pathRegexp = /^path (includes|does not include) (.*)/;
    private readonly descriptionRegexp =
        /^description (includes|does not include) (.*)/;
    private readonly sortByRegexp =
        /^sort by (status|due|done|path|description)/;

    private readonly headingRegexp =
        /^heading (includes|does not include) (.*)/;

    private readonly hideOptionsRegexp =
        /^hide (task count|backlink|done date|due date|recurrence rule|edit button)/;

    private readonly recurringString = 'is recurring';
    private readonly notRecurringString = 'is not recurring';

    private readonly limitRegexp = /^limit (to )?(\d+)( tasks?)?/;
    private readonly excludeSubItemsString = 'exclude sub-items';

    constructor({ source }: { source: string }) {
        source
            .split('\n')
            .map((line: string) => line.trim())
            .forEach((line: string) => {
                switch (true) {
                    case line === '':
                        break;
                    case line === this.doneString:
                        this._filters.push(
                            (task) => task.status === Status.Done,
                        );
                        break;
                    case line === this.notDoneString:
                        this._filters.push(
                            (task) => task.status !== Status.Done,
                        );
                        break;
                    case line === this.recurringString:
                        this._filters.push(
                            (task) => task.recurrenceRule !== null,
                        );
                        break;
                    case line === this.notRecurringString:
                        this._filters.push(
                            (task) => task.recurrenceRule === null,
                        );
                        break;
                    case line === this.excludeSubItemsString:
                        this._filters.push((task) => task.indentation === '');
                        break;
                    case line === this.noDueString:
                        this._filters.push((task) => task.dueDate === null);
                        break;
                    case this.dueRegexp.test(line):
                        this.parseDueFilter({ line });
                        break;
                    case this.doneRegexp.test(line):
                        this.parseDoneFilter({ line });
                        break;
                    case this.pathRegexp.test(line):
                        this.parsePathFilter({ line });
                        break;
                    case this.descriptionRegexp.test(line):
                        this.parseDescriptionFilter({ line });
                        break;
                    case this.headingRegexp.test(line):
                        this.parseHeadingFilter({ line });
                        break;
                    case this.limitRegexp.test(line):
                        this.parseLimit({ line });
                        break;
                    case this.sortByRegexp.test(line):
                        this.parseSortBy({ line });
                        break;
                    case this.hideOptionsRegexp.test(line):
                        this.parseHideOptions({ line });
                        break;
                    default:
                        this._error = 'do not understand query';
                }
            });
    }

    public get limit(): number | undefined {
        return this._limit;
    }

    public get layoutOptions(): LayoutOptions {
        return this._layoutOptions;
    }

    public get filters(): ((task: Task) => boolean)[] {
        return this._filters;
    }

    public get sorting(): Sorting[] {
        return this._sorting;
    }

    public get error(): string | undefined {
        return this._error;
    }

    private parseHideOptions({ line }: { line: string }): void {
        const hideOptionsMatch = line.match(this.hideOptionsRegexp);
        if (hideOptionsMatch !== null) {
            const option = hideOptionsMatch[1].trim().toLowerCase();

            if (option === 'task count') {
                this._layoutOptions.hideTaskCount = true;
            } else if (option === 'backlink') {
                this._layoutOptions.hideBacklinks = true;
            } else if (option === 'done date') {
                this._layoutOptions.hideDoneDate = true;
            } else if (option === 'due date') {
                this._layoutOptions.hideDueDate = true;
            } else if (option === 'recurrence rule') {
                this._layoutOptions.hideRecurrenceRule = true;
            } else if (option === 'edit button') {
                this._layoutOptions.hideEditButton = true;
            } else {
                this._error = 'do not understand hide option';
            }
        }
    }

    private parseDueFilter({ line }: { line: string }): void {
        const dueMatch = line.match(this.dueRegexp);
        if (dueMatch !== null) {
            const filterDate = this.parseDate(dueMatch[2]);
            if (!filterDate.isValid()) {
                this._error = 'do not understand due date';
            }

            let filter;
            if (dueMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isBefore(filterDate) : false;
            } else if (dueMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isSame(filterDate) : false;
            }

            this._filters.push(filter);
        } else {
            this._error = 'do not understand query filter (due date)';
        }
    }

    private parseDoneFilter({ line }: { line: string }): void {
        const doneMatch = line.match(this.doneRegexp);
        if (doneMatch !== null) {
            const filterDate = this.parseDate(doneMatch[2]);
            if (!filterDate.isValid()) {
                this._error = 'do not understand done date';
            }

            let filter;
            if (doneMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isBefore(filterDate) : false;
            } else if (doneMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isSame(filterDate) : false;
            }

            this._filters.push(filter);
        }
    }

    private parsePathFilter({ line }: { line: string }): void {
        const pathMatch = line.match(this.pathRegexp);
        if (pathMatch !== null) {
            const filterMethod = pathMatch[1];
            if (filterMethod === 'includes') {
                this._filters.push((task: Task) =>
                    task.path.includes(pathMatch[2]),
                );
            } else if (pathMatch[1] === 'does not include') {
                this._filters.push(
                    (task: Task) => !task.path.includes(pathMatch[2]),
                );
            } else {
                this._error = 'do not understand query filter (path)';
            }
        } else {
            this._error = 'do not understand query filter (path)';
        }
    }

    private parseDescriptionFilter({ line }: { line: string }): void {
        const descriptionMatch = line.match(this.descriptionRegexp);
        if (descriptionMatch !== null) {
            const filterMethod = descriptionMatch[1];
            if (filterMethod === 'includes') {
                this._filters.push((task: Task) =>
                    this.stringIncludesCaseInsensitive(
                        task.description,
                        descriptionMatch[2],
                    ),
                );
            } else if (descriptionMatch[1] === 'does not include') {
                this._filters.push(
                    (task: Task) =>
                        !this.stringIncludesCaseInsensitive(
                            task.description,
                            descriptionMatch[2],
                        ),
                );
            } else {
                this._error = 'do not understand query filter (description)';
            }
        } else {
            this._error = 'do not understand query filter (description)';
        }
    }

    private parseHeadingFilter({ line }: { line: string }): void {
        const headingMatch = line.match(this.headingRegexp);
        if (headingMatch !== null) {
            const filterMethod = headingMatch[1].toLowerCase();
            if (filterMethod === 'includes') {
                this._filters.push(
                    (task: Task) =>
                        task.precedingHeader !== null &&
                        this.stringIncludesCaseInsensitive(
                            task.precedingHeader,
                            headingMatch[2],
                        ),
                );
            } else if (headingMatch[1] === 'does not include') {
                this._filters.push(
                    (task: Task) =>
                        task.precedingHeader !== null &&
                        !this.stringIncludesCaseInsensitive(
                            task.precedingHeader,
                            headingMatch[2],
                        ),
                );
            } else {
                this._error = 'do not understand query filter (heading)';
            }
        } else {
            this._error = 'do not understand query filter (heading)';
        }
    }

    private parseLimit({ line }: { line: string }): void {
        const limitMatch = line.match(this.limitRegexp);
        if (limitMatch !== null) {
            // limitMatch[2] is per regex always digits and therefore parsable.
            const limit = Number.parseInt(limitMatch[2], 10);
            this._limit = limit;
        } else {
            this._error = 'do not understand query limit';
        }
    }

    private parseSortBy({ line }: { line: string }): void {
        const fieldMatch = line.match(this.sortByRegexp);
        if (fieldMatch !== null) {
            this._sorting.push(fieldMatch[1] as Sorting);
        } else {
            this._error = 'do not understand query sorting';
        }
    }

    private parseDate(input: string): moment.Moment {
        // Using start of date to correctly match on comparison with other dates (like equality).
        return window.moment(chrono.parseDate(input)).startOf('day');
    }

    private stringIncludesCaseInsensitive(
        haystack: string,
        needle: string,
    ): boolean {
        return haystack
            .toLocaleLowerCase()
            .includes(needle.toLocaleLowerCase());
    }
}
