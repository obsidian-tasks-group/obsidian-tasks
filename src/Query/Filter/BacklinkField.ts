import type { Task } from '../../Task';
import type { GrouperFunction } from '../Grouper';
import { TextField } from './TextField';
import { FilterOrErrorMessage } from './Filter';

export class BacklinkField extends TextField {
    public fieldName(): string {
        return 'backlink';
    }

    public value(task: Task): string {
        const linkText = task.getLinkText({ isFilenameUnique: true });
        if (linkText === null) {
            return 'Unknown Location';
        }
        return linkText;
    }

    createFilterOrErrorMessage(line: string): FilterOrErrorMessage {
        return FilterOrErrorMessage.fromError(line, 'backlink field does not support filtering');
    }

    canCreateFilterForLine(_line: string): boolean {
        return false;
    }

    public supportsGrouping(): boolean {
        return true;
    }

    public grouper(): GrouperFunction {
        return (task: Task) => {
            const filename = task.filename;
            if (filename === null) {
                return ['Unknown Location'];
            }

            // Always escape the filename, to prevent any underscores being interpreted as markdown:
            let result = TextField.escapeMarkdownCharacters(filename);

            // Only append the heading if it differs from the un-escaped fileanme:
            if (task.precedingHeader && task.precedingHeader !== filename) {
                // We don't escape the heading, as any markdown characters in it really are markdown styling:
                result += ' > ' + task.precedingHeader;
            }
            return [result];
        };
    }
}
