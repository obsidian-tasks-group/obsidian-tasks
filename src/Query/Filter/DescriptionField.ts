import { getSettings } from '../../config/Settings';
import type { Task } from '../../Task';
import { TextField } from './TextField';

/**
 * Support the 'description' search instruction.
 *
 * Note that DescriptionField.value() returns the description
 * with the global filter (if any) removed.
 */
export class DescriptionField extends TextField {
    protected fieldName(): string {
        return 'description';
    }

    /**
     * Return the task's description, with any global tag removed
     *
     * Promoted to public, to enable testing.
     * @param task
     * @public
     */
    public value(task: Task): string {
        // Remove global filter from description match if present.
        // This is necessary to match only on the content of the task, not
        // the global filter.
        const globalFilter = getSettings().globalFilter;
        return task.description.replace(globalFilter, '').trim();
    }
}
