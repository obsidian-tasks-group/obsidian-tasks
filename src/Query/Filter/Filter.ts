import type { Task } from '../../Task';

/**
 * A filtering function, that takes a Task object and returns
 * whether it matches a particular filtering instruction.
 */
export type Filter = (task: Task) => boolean;

/**
 * A class which stores one of:
 * - A Filter
 * - An error message
 *
 * This is really currently a convenience for returning date from
 * Field.createFilterOrErrorMessage() and derived classes.
 *
 * Later, it may gain helper functions for constructing parser error messages,
 * as currently these are created by some rather repetitious code, and also
 * there is scope for making these messages more informative (including the
 * problem line, and perhaps listing allowed options).
 */
export class FilterOrErrorMessage {
    filter: Filter | undefined;
    error: string | undefined;

    /**
     * Construct a FilterOrErrorMessage with the filter.
     * @param filter
     */
    public static fromFilter(filter: Filter): FilterOrErrorMessage {
        const result = new FilterOrErrorMessage();
        result.filter = filter;
        return result;
    }

    /**
     * Construct a FilterOrErrorMessage with the given error message.
     * @param errorMessage
     */
    public static fromError(errorMessage: string): FilterOrErrorMessage {
        const result = new FilterOrErrorMessage();
        result.error = errorMessage;
        return result;
    }
}
