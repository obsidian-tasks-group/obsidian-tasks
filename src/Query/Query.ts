import { LayoutOptions } from '../TaskLayout';
import type { Task } from '../Task';
import type { IQuery } from '../IQuery';
import { getSettings } from '../Config/Settings';
import { Sort } from './Sort';
import type { Sorter } from './Sorter';
import { TaskGroups } from './TaskGroups';
import * as FilterParser from './FilterParser';
import type { Grouper } from './Grouper';
import type { Filter } from './Filter/Filter';

export class Query implements IQuery {
    public source: string;

    private _limit: number | undefined = undefined;
    private _taskGroupLimit: number | undefined = undefined;
    private _layoutOptions: LayoutOptions = new LayoutOptions();
    private _filters: Filter[] = [];
    private _error: string | undefined = undefined;
    private _sorting: Sorter[] = [];
    private _grouping: Grouper[] = [];

    private readonly hideOptionsRegexp =
        /^(hide|show) (task count|backlink|priority|created date|start date|scheduled date|done date|due date|recurrence rule|edit button|urgency)/;
    private readonly shortModeRegexp = /^short/;
    private readonly explainQueryRegexp = /^explain/;

    private readonly limitRegexp = /^limit (groups )?(to )?(\d+)( tasks?)?/;

    private readonly commentRegexp = /^#.*/;

    constructor({ source }: { source: string }) {
        this.source = source;
        source
            .split('\n')
            .map((line: string) => line.trim())
            .forEach((line: string) => {
                switch (true) {
                    case line === '':
                        break;
                    case this.shortModeRegexp.test(line):
                        this._layoutOptions.shortMode = true;
                        break;
                    case this.explainQueryRegexp.test(line):
                        this._layoutOptions.explainQuery = true;
                        break;
                    case this.limitRegexp.test(line):
                        this.parseLimit({ line });
                        break;
                    case this.parseSortBy({ line }):
                        break;
                    case this.parseGroupBy({ line }):
                        break;
                    case this.hideOptionsRegexp.test(line):
                        this.parseHideOptions({ line });
                        break;
                    case this.commentRegexp.test(line):
                        // Comment lines are ignored
                        break;
                    case this.parseFilter(line):
                        break;
                    default:
                        this._error = `do not understand query: ${line}`;
                }
            });
    }

    /**
     *
     * Appends {@link q2} to this query.
     *
     * @note At time of writing, this query language appears to play nicely with combining queries.
     *
     * More formally, the concatenation operation on the query language:
     *     * Is closed (concatenating two queries is another valid query)
     *     * Is not commutative (q1.append(q2) !== q2.append(q1))
     *
     * And the semantics of the combination are straight forward:
     *     * Combining two queries appends their filters
     *           (assuming that the filters are pure functions, filter concatenation is commutative)
     *     * Combining two queries appends their sorting instructions. (this is not commutative)
     *     * Combining two queries appends their grouping instructions. (this is not commutative)
     *     * Successive limit instructions overwrite previous ones.
     *
     * @param {Query} q2
     * @return {Query} The combined query
     */
    public append(q2: Query): Query {
        if (this.source === '') return q2;
        if (q2.source === '') return this;
        return new Query({ source: `${this.source}\n${q2.source}` });
    }

    /**
     * Generate a text description of the contents of this query.
     *
     * This does not show any global filter and global query.
     * Use {@link explainResults} if you want to see any global query and global filter as well.
     */
    public explainQuery(): string {
        let result = '';

        const numberOfFilters = this.filters.length;
        if (numberOfFilters === 0) {
            result += 'No filters supplied. All tasks will match the query.';
        } else {
            for (let i = 0; i < numberOfFilters; i++) {
                if (i > 0) result += '\n';
                result += this.filters[i].explainFilterIndented('');
            }
        }
        result += this.explainQueryLimits();

        const { debugSettings } = getSettings();
        if (debugSettings.ignoreSortInstructions) {
            result +=
                "\n\nNOTE: All sort instructions, including default sort order, are disabled, due to 'ignoreSortInstructions' setting.";
        }

        return result;
    }

    private explainQueryLimits() {
        let result = '';

        function getPluralisedText(limit: number) {
            let text = `\n\nAt most ${limit} task`;
            if (limit !== 1) {
                text += 's';
            }
            return text;
        }

        if (this._limit !== undefined) {
            result += getPluralisedText(this._limit);
            result += '.\n';
        }

        if (this._taskGroupLimit !== undefined) {
            result += getPluralisedText(this._taskGroupLimit);
            result += ' per group (if any "group by" options are supplied).\n';
        }
        return result;
    }

    public get limit(): number | undefined {
        return this._limit;
    }

    public get layoutOptions(): LayoutOptions {
        return this._layoutOptions;
    }

    public get filters(): Filter[] {
        return this._filters;
    }

    public get sorting() {
        return this._sorting;
    }

    /**
     * Return the {@link Grouper} objects that represent any `group by` instructions in the tasks block.
     */
    public get grouping(): Grouper[] {
        return this._grouping;
    }

    public get error(): string | undefined {
        return this._error;
    }

    public applyQueryToTasks(tasks: Task[]): TaskGroups {
        this.filters.forEach((filter) => {
            tasks = tasks.filter(filter.filterFunction);
        });

        const { debugSettings } = getSettings();
        const tasksSorted = debugSettings.ignoreSortInstructions ? tasks : Sort.by(this.sorting, tasks);
        const tasksSortedLimited = tasksSorted.slice(0, this.limit);

        const taskGroups = new TaskGroups(this.grouping, tasksSortedLimited);

        if (this._taskGroupLimit !== undefined) {
            taskGroups.applyTaskLimit(this._taskGroupLimit);
        }

        return taskGroups;
    }

    private parseHideOptions({ line }: { line: string }): void {
        const hideOptionsMatch = line.match(this.hideOptionsRegexp);
        if (hideOptionsMatch !== null) {
            const hide = hideOptionsMatch[1] === 'hide';
            const option = hideOptionsMatch[2];

            switch (option) {
                case 'task count':
                    this._layoutOptions.hideTaskCount = hide;
                    break;
                case 'backlink':
                    this._layoutOptions.hideBacklinks = hide;
                    break;
                case 'priority':
                    this._layoutOptions.hidePriority = hide;
                    break;
                case 'created date':
                    this._layoutOptions.hideCreatedDate = hide;
                    break;
                case 'start date':
                    this._layoutOptions.hideStartDate = hide;
                    break;
                case 'scheduled date':
                    this._layoutOptions.hideScheduledDate = hide;
                    break;
                case 'due date':
                    this._layoutOptions.hideDueDate = hide;
                    break;
                case 'done date':
                    this._layoutOptions.hideDoneDate = hide;
                    break;
                case 'recurrence rule':
                    this._layoutOptions.hideRecurrenceRule = hide;
                    break;
                case 'edit button':
                    this._layoutOptions.hideEditButton = hide;
                    break;
                case 'urgency':
                    this._layoutOptions.hideUrgency = hide;
                    break;
                default:
                    this._error = 'do not understand hide/show option';
            }
        }
    }

    private parseFilter(line: string) {
        const filterOrError = FilterParser.parseFilter(line);
        if (filterOrError != null) {
            if (filterOrError.filter) this._filters.push(filterOrError.filter);
            else this._error = filterOrError.error;
            return true;
        }
        return false;
    }

    private parseLimit({ line }: { line: string }): void {
        const limitMatch = line.match(this.limitRegexp);
        if (limitMatch === null) {
            this._error = 'do not understand query limit';
            return;
        }

        // limitMatch[3] is per regex always digits and therefore parsable.
        const limitFromLine = Number.parseInt(limitMatch[3], 10);

        if (limitMatch[1] !== undefined) {
            this._taskGroupLimit = limitFromLine;
        } else {
            this._limit = limitFromLine;
        }
    }

    private parseSortBy({ line }: { line: string }): boolean {
        const sortingMaybe = FilterParser.parseSorter(line);
        if (sortingMaybe) {
            this._sorting.push(sortingMaybe);
            return true;
        }
        return false;
    }

    /**
     * Parsing of `group by` lines, for grouping that is implemented in the {@link Field}
     * classes.
     *
     * @param line
     * @private
     */
    private parseGroupBy({ line }: { line: string }): boolean {
        const groupingMaybe = FilterParser.parseGrouper(line);
        if (groupingMaybe) {
            this._grouping.push(groupingMaybe);
            return true;
        }
        return false;
    }
}
