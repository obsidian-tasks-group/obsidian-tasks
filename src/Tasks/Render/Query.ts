import moment from 'moment';
import { Status } from 'Tasks/Status';
import { DATE_FORMAT, Task } from '../Task';

export class Query {
    public readonly filters: ((task: Task) => boolean)[];
    private readonly noDueString = 'no due date';
    private readonly dueRegexp = /due (before|after|on) (\d{4}-\d{2}-\d{2})/;
    private readonly doneString = 'done';
    private readonly notDoneString = 'not done';
    private readonly doneRegexp = /done (before|after|on) (\d{4}-\d{2}-\d{2})/;
    private readonly pathRegexp = /path (includes|does not include) (.*)/;
    private readonly excludeSubItemsString = 'exclude sub-items';

    constructor({ source }: { source: string }) {
        this.filters = this.parseFilters({ source });
    }

    private parseFilters({
        source,
    }: {
        source: string;
    }): ((task: Task) => boolean)[] {
        const filters: ((task: Task) => boolean)[] = [];
        const sourceLines = source.split('\n').map(line => line.trim());

        for (const sourceLine of sourceLines) {
            if (sourceLine === '') {
                continue;
            }

            if (sourceLine === this.doneString) {
                filters.push((task) => task.status === Status.DONE);
            } else if (sourceLine === this.notDoneString) {
                filters.push((task) => task.status !== Status.DONE);
            } else if (sourceLine === this.noDueString) {
                filters.push((task) => task.dueDate === undefined);
            } else if (this.dueRegexp.test(sourceLine)) {
                const filter = this.parseDueFilter({ line: sourceLine });
                if (filter !== undefined) {
                    filters.push(filter);
                }
            } else if (this.doneRegexp.test(sourceLine)) {
                const filter = this.parseDoneFilter({ line: sourceLine });
                if (filter !== undefined) {
                    filters.push(filter);
                }
            } else if (this.pathRegexp.test(sourceLine)) {
                const filter = this.parsePathFilter({ line: sourceLine });
                if (filter !== undefined) {
                    filters.push(filter);
                }
            } else if (sourceLine === this.excludeSubItemsString) {
                filters.push((task) => task.indentation === '');
            } else {
                console.error('Tasks: do not understand filter: ' + sourceLine);
            }
        }

        return filters;
    }

    private parseDueFilter({
        line,
    }: {
        line: string;
    }): ((task: Task) => boolean) | undefined {
        const dueMatch = line.match(this.dueRegexp);
        if (dueMatch !== null) {
            let filter;
            const filterDate = moment(dueMatch[2], DATE_FORMAT);
            if (dueMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isBefore(filterDate) : false;
            } else if (dueMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isSame(filterDate) : false;
            }

            return filter;
        }

        return undefined;
    }

    private parseDoneFilter({
        line,
    }: {
        line: string;
    }): ((task: Task) => boolean) | undefined {
        const doneMatch = line.match(this.doneRegexp);
        if (doneMatch !== null) {
            let filter;
            const filterDate = moment(doneMatch[2], DATE_FORMAT);
            if (doneMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isBefore(filterDate) : false;
            } else if (doneMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isSame(filterDate) : false;
            }

            return filter;
        }

        return undefined;
    }

    private parsePathFilter({
        line,
    }: {
        line: string;
    }): ((task: Task) => boolean) | undefined {
        const pathMatch = line.match(this.pathRegexp);
        if (pathMatch !== null) {
            let filter;
            const filterMethod = pathMatch[1];
            if (filterMethod === 'includes') {
                filter = (task: Task) =>
                    task.path.includes(pathMatch[2]);
            } else if (pathMatch[1] === 'does not include') {
                filter = (task: Task) =>
                    !task.path.includes(pathMatch[2]);
            } else {
                console.error('Tasks: do not understand path query: ' + line);
                return undefined;
            }

            return filter;
        }

        return undefined;
    }
}
