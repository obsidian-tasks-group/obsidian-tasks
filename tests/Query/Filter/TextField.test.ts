import { verify } from 'approvals/lib/Providers/Jest/JestApprovals';
import { DescriptionField } from '../../../src/Query/Filter/DescriptionField';
import { Query } from '../../../src/Query/Query';

describe('should report regular expression errors to user ', () => {
    it('should report mismatched parenthesis', () => {
        const instruction = 'description regex matches /hello(/';
        const filterOrError = new DescriptionField().createFilterOrErrorMessage(instruction);
        expect(filterOrError.error).toEqual(
            'Error: Parsing regular expression. The error message was: "SyntaxError: Invalid regular expression: /hello(/: Unterminated group"',
        );
    });
});

describe('explains regular sub-string searches', () => {
    it('should explain simple string search', () => {
        const instruction = 'description includes hello';
        const field = new DescriptionField().createFilterOrErrorMessage(instruction);
        expect(field).toHaveExplanation('description includes hello');
    });
});

describe('explains regular expression searches', () => {
    it('should explain regex matches search', () => {
        const instruction = 'description regex matches /hello/';
        const field = new DescriptionField().createFilterOrErrorMessage(instruction);
        expect(field).toHaveExplanation("using regex:            'hello' with no flags");
    });

    it('should explain regex does not match search', () => {
        const instruction = 'description regex does not match /hello/';
        const field = new DescriptionField().createFilterOrErrorMessage(instruction);
        expect(field).toHaveExplanation("using regex:                   'hello' with no flags");
    });

    it('bulk test', () => {
        /* This test uses Approval Tests' verify() to quickly get good coverage on a wide variety
           of regular expression examples.
         */

        // Arrange
        // Alphabetical order please
        const source = `
            description regex does not match /(buy|order|voucher|lakeland|purchase|\\spresent)/i
            description regex does not match /^$/
            description regex matches /  /
            description regex matches /#context/pc_photos|#context/pc_clare|#context/pc_macbook/i
            description regex matches /#context\\/pc_photos|#context\\/pc_clare|#context\\/pc_macbook/i
            description regex matches /#tag\\/subtag[0-9]\\/subsubtag[0-9]/i
            description regex matches /(buy|order|voucher|lakeland|purchase|\\spresent)/i
            description regex matches /[⏫🔼🔽📅⏳🛫🔁]/
            description regex matches /[⏫🔼🔽📅⏳🛫🔁]/u
            description regex matches /^$/
            description regex matches /^Log/i
            description regex matches /waiting|waits|wartet/
            description regex matches /waiting|waits|wartet/i
            filename regex does not match /^Tasks User Support Kanban\\.md$/
            folder regex matches /root/sub-folder/sub-sub-folder/
            folder regex matches /root\\/sub-folder\\/sub-sub-folder/
            heading regex matches /^Exactly Matched Heading$/
            path regex does not match /^_meta/
            path regex does not match /^_templates/
            path regex does not match /sadfasfdafa/i
            path regex does not match /sadfasfdafa/igm
            path regex does not match /w.bble/
            path regex matches /(george phone|exercise|100th|knee|TRVs|HWRC|2023-03-21 Python Pairing on the Hub with so-and-so|Epson|Fred's flat|1519 - feat - Theme-ability|McDermid)/i
            path regex matches /(george)/i
            path regex matches /Log/
            path regex matches /clare/i
            path regex matches /root/sub-folder/sub-sub-folder/index\\.md/
            path regex matches /root\\/sub-folder\\/sub-sub-folder\\/index\\.md/
            path regex matches /sadfasfdafa/i
            recurrence regex does not match /\\d/
            recurrence regex matches /\\d/
            tag regex does not match /./
            tag regex matches /#book$/
            tag regex matches /#t$/
            tags regex does not match /(home|town)/
            tags regex matches /(home|pc_mac|town)/
`;
        const query = new Query({ source });

        // Assert
        verify(query.explainQuery() + `\nError: ${query.error ?? 'None'}`);
    });
});
