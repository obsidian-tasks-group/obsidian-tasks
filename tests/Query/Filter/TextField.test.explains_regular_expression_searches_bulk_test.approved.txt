description regex does not match /(buy|order|voucher|lakeland|purchase|\spresent)/i =>
  using regex:                   '(buy|order|voucher|lakeland|purchase|\spresent)' with flag 'i'

description regex does not match /^$/ =>
  using regex:                   '^$' with no flags

description regex matches /  / =>
  using regex:            '  ' with no flags

description regex matches /#context/pc_photos|#context/pc_clare|#context/pc_macbook/i =>
  using regex:            '#context' with no flags

description regex matches /#context\/pc_photos|#context\/pc_clare|#context\/pc_macbook/i =>
  using regex:            '#context\/pc_photos|#context\/pc_clare|#context\/pc_macbook' with flag 'i'

description regex matches /#tag\/subtag[0-9]\/subsubtag[0-9]/i =>
  using regex:            '#tag\/subtag[0-9]\/subsubtag[0-9]' with flag 'i'

description regex matches /(buy|order|voucher|lakeland|purchase|\spresent)/i =>
  using regex:            '(buy|order|voucher|lakeland|purchase|\spresent)' with flag 'i'

description regex matches /[⏫🔼🔽📅⏳🛫🔁]/ =>
  using regex:            '[⏫🔼🔽📅⏳🛫🔁]' with no flags

description regex matches /[⏫🔼🔽📅⏳🛫🔁]/u =>
  using regex:            '[⏫🔼🔽📅⏳🛫🔁]' with no flags

description regex matches /^$/ =>
  using regex:            '^$' with no flags

description regex matches /^Log/i =>
  using regex:            '^Log' with flag 'i'

description regex matches /waiting|waits|wartet/ =>
  using regex:            'waiting|waits|wartet' with no flags

description regex matches /waiting|waits|wartet/i =>
  using regex:            'waiting|waits|wartet' with flag 'i'

filename regex does not match /^Tasks User Support Kanban\.md$/ =>
  using regex:                '^Tasks User Support Kanban\.md$' with no flags

folder regex matches /root/sub-folder/sub-sub-folder/ =>
  using regex:       'root' with no flags

folder regex matches /root\/sub-folder\/sub-sub-folder/ =>
  using regex:       'root\/sub-folder\/sub-sub-folder' with no flags

heading regex matches /^Exactly Matched Heading$/ =>
  using regex:        '^Exactly Matched Heading$' with no flags

path regex does not match /^_meta/ =>
  using regex:            '^_meta' with no flags

path regex does not match /^_templates/ =>
  using regex:            '^_templates' with no flags

path regex does not match /sadfasfdafa/i =>
  using regex:            'sadfasfdafa' with flag 'i'

path regex does not match /sadfasfdafa/igm =>
  using regex:            'sadfasfdafa' with flags 'gim'

path regex does not match /w.bble/ =>
  using regex:            'w.bble' with no flags

path regex matches /(george phone|exercise|100th|knee|TRVs|HWRC|2023-03-21 Python Pairing on the Hub with so-and-so|Epson|Fred's flat|1519 - feat - Theme-ability|McDermid)/i =>
  using regex:     '(george phone|exercise|100th|knee|TRVs|HWRC|2023-03-21 Python Pairing on the Hub with so-and-so|Epson|Fred's flat|1519 - feat - Theme-ability|McDermid)' with flag 'i'

path regex matches /(george)/i =>
  using regex:     '(george)' with flag 'i'

path regex matches /Log/ =>
  using regex:     'Log' with no flags

path regex matches /clare/i =>
  using regex:     'clare' with flag 'i'

path regex matches /root/sub-folder/sub-sub-folder/index\.md/ =>
  using regex:     'root' with no flags

path regex matches /root\/sub-folder\/sub-sub-folder\/index\.md/ =>
  using regex:     'root\/sub-folder\/sub-sub-folder\/index\.md' with no flags

path regex matches /sadfasfdafa/i =>
  using regex:     'sadfasfdafa' with flag 'i'

recurrence regex does not match /\d/ =>
  using regex:                  '\d' with no flags

recurrence regex matches /\d/ =>
  using regex:           '\d' with no flags

tag regex does not match /./ =>
  using regex:           '.' with no flags

tag regex matches /#book$/ =>
  using regex:    '#book$' with no flags

tag regex matches /#t$/ =>
  using regex:    '#t$' with no flags

tags regex does not match /(home|town)/ =>
  using regex:            '(home|town)' with no flags

tags regex matches /(home|pc_mac|town)/ =>
  using regex:     '(home|pc_mac|town)' with no flags

Error: None
